File: src/main/java/algo/binarySearch/Heaters_475.java
package algo.binarySearch;

import java.util.Arrays;

public class Heaters_475 {
    public static int findRadius(int[] houses, int[] heaters) {
        Arrays.sort(houses);
        Arrays.sort(heaters);

        var radius = 0;
        for (int house : houses) {
          radius = Math.max(radius, search(house, heaters));
        }
        return radius;
    }

    public static int search(int house, int[] heaters) {
        if (house <= heaters[0]) {
            return heaters[0] - house;
        }
        if (house >= heaters[heaters.length - 1]) {
            return house - heaters[heaters.length - 1];
        }
        int start = 0, end = heaters.length - 1;
        while (start < end) {
            var mid = start + (end - start) / 2;
            if (mid == start) {
                return Math.min(house - heaters[start], heaters[end] - house);
            }
            if (heaters[mid] == house) {
                return 0;
            } else if (heaters[mid] < house) {
                start = mid;
            } else {
                end = mid;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        System.out.println(findRadius(new int[]{1, 5}, new int[]{10}));
        System.out.println(findRadius(new int[]{1,2,3,4}, new int[]{1,4}));
        System.out.println(findRadius(new int[]{25921153,510616708}, new int[]{771515668,357571490,44788124,927702196,952509530}));
    }
}

File: src/main/kotlin/algo/binarySearch/Heaters_475.kt
package algo.binarySearch

class KtHeaters_475 {
    fun findRadius(houses: IntArray, heaters: IntArray): Int {
        // for every house, binary search in heaters to locate the position of the house in heaters
        // and find the nearest heater for the house
        var radius = 0
        houses.sort()
        heaters.sort()
        houses.forEach { house ->
            radius = radius.coerceAtLeast(search(house, heaters))
        }
        return radius
    }

    // find the nearest heater for given house
    private fun search(house: Int, heaters: IntArray): Int {
        var start = 0; var end = heaters.size - 1
        if (house <= heaters[0]) {
            return heaters[0] - house
        }
        if (house >= heaters.last()) {
            return house - heaters.last()
        }
        while (start < end) {
            val mid = start + (end - start) / 2
            if (mid == start) {
                return (house - heaters[start]).coerceAtMost(heaters[end] - house)
            }
            if (heaters[mid] == house) {
                return 0
            } else if (heaters[mid] < house) {
                start = mid
            } else {
                end = mid
            }
        }
        return -1
    }
}

fun main() {
    val heater = KtHeaters_475()
    println(heater.findRadius(intArrayOf(1,2,3), intArrayOf(2)))
    println(heater.findRadius(intArrayOf(1,2,3,4), intArrayOf(1,4)))
    println(heater.findRadius(intArrayOf(1,5), intArrayOf(2)))
}